lib/bitcoin/BitcoinClient.ts

```diff
@@ -313,6 +325,156 @@ export default class BitcoinClient {
+   * @param paytoAddress The address where the spend transaction should go to.
+   */
+  private async createSpendBitcoreTransactionFromFrozenTransaction (
+    previousFreezeTransaction: Transaction,
+    previousFreezeUntilBlock: number,
+    paytoAddress: Address): Promise<Transaction> {
+
+    // First create an input from the previous frozen transaction output.
+    const frozenOutputAsInput = this.createBitcoreUnspentOutputFromFrozenTransaction(previousFreezeTransaction, previousFreezeUntilBlock);
+    const previousFreezeAmountInSatoshis = frozenOutputAsInput.satoshis;
+
+    // Now create a spend transaction using the frozen output. Create the transaction with all
+    // inputs and outputs as they are needed to calculate the fee.
+    const spendTransaction = new Transaction()
+                                   .from([frozenOutputAsInput])
+                                   .to(paytoAddress, previousFreezeAmountInSatoshis)
+                                   .lockUntilBlockHeight(previousFreezeUntilBlock);
+
+    const transactionFee = await this.calculateTransactionFee(spendTransaction);
+
+    // We need to set the transaction fee and subtract that fee from the freeze amount.
+    // We cannot just update the existing output (it's readonly), so we need to first remove it,
+    // and add another one with the correct amount.
+    spendTransaction.outputs.shift();
+    spendTransaction.to(paytoAddress, previousFreezeAmountInSatoshis - transactionFee)
+                    .fee(transactionFee);
+
+    // Now update the first input to include it's signature.
+    // Create an input in the format expected for a spend-from-freeze input and set it on the
+    // transaction.
+    const previousFreezeScript = this.createFreezeBitcoreScript(previousFreezeUntilBlock);
+    const signature = (Transaction as any).sighash.sign(spendTransaction, this.privateKey, 0x1, 0, previousFreezeScript);
+    const inputScript = Script.empty()
+                              .add(signature.toTxFormat())
+                              .add(this.privateKey.toPublicKey().toBuffer())
+                              .add(previousFreezeScript.toBuffer());
+
+    (spendTransaction.inputs[0] as any).setScript(inputScript);
+
+    return spendTransaction;
+  }
+
+  private createBitcoreUnspentOutputFromFrozenTransaction (
+    previousFreezeTransaction: Transaction,
+    previousFreezeUntilBlock: number): Transaction.UnspentOutput {
+
+    const previousFreezeAmountInSatoshis = previousFreezeTransaction.outputs[0].satoshis;
+    const previousFreezeRedeemScript = this.createFreezeBitcoreScript(previousFreezeUntilBlock);
+    const previousFreezeRedeemScriptHash = Script.buildScriptHashOut(previousFreezeRedeemScript);
+
+    const frozenOutputAsUnspentOutput = Transaction.UnspentOutput.fromObject({
+      txid: previousFreezeTransaction.id,
+      vout: 0,
+      scriptPubKey: previousFreezeRedeemScriptHash,
+      satoshis: previousFreezeAmountInSatoshis

```

Comment:
For my understanding: I was under the impression prior to this review that there is no such thing as spending a fraction of an unspent transaction output, thus surprise that you can actually specify satoshi amount? What happens if one specify amount less than total available?

Reply:
If you specify the amount less than the total then the rest goes to the miner. I think that the library requires this so that it can do validations on the input/output/fee calculations.